pinot-tools/src/main/java/org/apache/pinot/tools/anonymizer/MapBasedGlobalDictionaries.java

```diff
@@ -0,0 +1,278 @@
+   * dictionaries by generating values:
+   *
+   * For numeric columns, we generate in order since
+   * we start with a base value.
+   * For string column, we first generate and then sort
+   */
+  @Override
+  public void addDerivedValuesToGlobalDictionaries() {
+    // update global dictionary for each column by adding
+    // the corresponding generated value for each orig value
+    for (Map.Entry<String, OrigAndDerivedValueHolder> entry : _columnToGlobalDictionary.entrySet()) {
+      OrigAndDerivedValueHolder origAndDerivedValueHolder = entry.getValue();
+      generateDerivedValuesForGlobalDictionary(origAndDerivedValueHolder);
+    }
+  }
+
+  @Override
+  public void serialize(String outputDir) throws Exception {
+    // write global dictionary for each column
+    for (String column : _columnToGlobalDictionary.keySet()) {
+      PrintWriter dictionaryWriter = new PrintWriter(new BufferedWriter(new FileWriter(outputDir + "/" + column + DICT_FILE_EXTENSION)));
+      OrigAndDerivedValueHolder origAndDerivedValueHolder = _columnToGlobalDictionary.get(column);
+      Set<Map.Entry<Object, DerivedValue>> entries = origAndDerivedValueHolder._origAndDerivedValues.entrySet();
+      Iterator<Map.Entry<Object, DerivedValue>> sortedIterator = entries.iterator();
+      while (sortedIterator.hasNext()) {
+        Map.Entry<Object, DerivedValue> entry = sortedIterator.next();
+        dictionaryWriter.println(entry.getKey());
+        dictionaryWriter.println(entry.getValue()._derivedValue);
+      }
+      dictionaryWriter.flush();
+    }
+  }
+
+  @Override
+  public Object getDerivedValueForOrigValueSV(String column, Object origValue) {
+    OrigAndDerivedValueHolder origAndDerivedValueHolder = _columnToGlobalDictionary.get(column);
+    TreeMap<Object, DerivedValue> sortedMap = origAndDerivedValueHolder._origAndDerivedValues;
+    return sortedMap.get(origValue)._derivedValue;
+  }
+
+  @Override
+  public Object[] getDerivedValuesForOrigValuesMV(String column, Object[] origMultiValues) {
+    OrigAndDerivedValueHolder origAndDerivedValueHolder = _columnToGlobalDictionary.get(column);
+    TreeMap<Object, DerivedValue> sortedMap = origAndDerivedValueHolder._origAndDerivedValues;
+    int length = origMultiValues.length;
+    Object[] derivedMultiValues = new Object[length];
+    for (int i = 0; i < length; i++) {
+      derivedMultiValues[i] = sortedMap.get(origMultiValues[i]);
+    }
+    return derivedMultiValues;
+  }
+
+  private static class OrigAndDerivedValueHolder {
+    FieldSpec.DataType _dataType;
+    TreeMap<Object, DerivedValue> _origAndDerivedValues;
+
+    OrigAndDerivedValueHolder(FieldSpec.DataType dataType) {
+      _dataType = dataType;
+      _origAndDerivedValues = new TreeMap<>();
+    }
+
+    void setOrigValue(Object origValue) {
+     if (!_origAndDerivedValues.containsKey(origValue)) {
+       _origAndDerivedValues.put(origValue, new DerivedValue());
+     }
+    }
+  }
+
+  private static class DerivedValue {
```

Comment:
This wrapper is needed so that we can set the derived values in global dictionary in sorted order while iterating over it

Reply:
Can you put this comment on top of this class?

Reply:
Done